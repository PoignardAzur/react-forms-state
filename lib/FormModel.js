// @flow

import { fromJS, List, Map } from "immutable";

export type ConversionModel = {
  out?: string,
  out_path?: string,
  default?: ?any,
  convertIn?: (value: any) => any,
  convertOut?: (value: any) => any,
  complex?: boolean,
  [key: string]: ConversionModel,
};

export type ConversionJob = {
  in: string,
  out: string,
  convertIn: (v: any) => any,
  convertOut: (v: any) => any,
  default: any,
};

export type ConversionJobs = List<ConversionJob>;

export function convertConversionModelToConversionJobs(
  model: ConversionModel,
  currentInPath?: string = "",
  parentOutPath?: string = "",
): ConversionJobs {
  if (typeof model !== "object" || model == null) {
    return List();
  }
  const currentNodeOutPath = "out" in model || "out_path" in model
    ? model.out || model.out_path
    : "";
  const currentOutPath = `${parentOutPath != ""
    ? `${parentOutPath}.${currentNodeOutPath}`
    : currentNodeOutPath}`;
  const jobs = List()
    .concat(
      currentInPath !== ""
        ? List.of({
            in: currentInPath,
            out: currentOutPath,
            convertIn: "convertIn" in model ? model.convertIn : value => value,
            convertOut: "convertOut" in model
              ? model.convertOut
              : value => value,
            default: model.default,
          })
        : List(),
    )
    .concat(
      Object.keys(model)
        .filter(
          nodeKey =>
            ![
              "convertIn",
              "convertOut",
              "out",
              "out_path",
              "complex",
              "default",
            ].includes(nodeKey),
        )
        .reduce((red: ConversionJobs, nodeKey: string): ConversionJobs => {
          const childNode = model[nodeKey];
          const childInPath = currentInPath
            ? `${currentInPath}.${nodeKey}`
            : nodeKey;
          return red.concat(
            convertConversionModelToConversionJobs(
              childNode,
              childInPath,
              currentOutPath,
            ),
          );
        }, List()),
    );
  return jobs;
}

export function convertIn(value?: ?Object, jobs: ConversionJobs): Object {
  if (value == null) return {};
  const immutableValue = fromJS(value);
  return jobs
    .reduceRight((red, job) => {
      const inValue = job.convertIn(
        immutableValue.getIn(job.in.split("."), job.default),
      );
      if (red.getIn(job.out.split(".")) != null && inValue != null) return red;
      return red.setIn(job.out.split("."), fromJS(inValue));
    }, Map())
    .toJS();
}

export function convertOut(
  value?: ?Object,
  jobs: ConversionJobs,
  props: any,
): Object {
  if (value == null) return {};
  const immutableValue = fromJS(value);
  return jobs
    .reduceRight((red, job) => {
      const outValue = job.convertOut(
        immutableValue.getIn(job.out.split(".")),
        props,
      );
      if (outValue === undefined) {
        return red;
      }
      if (red.getIn(job.in.split(".")) != null && outValue != null) return red;
      return red.setIn(job.in.split("."), outValue);
    }, Map())
    .toJS();
}
