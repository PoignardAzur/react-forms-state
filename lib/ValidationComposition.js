// @flow

import { fromJS } from "immutable";

export function notNull(
  { errorString = "Not filled" }: { errorString?: string } = {},
) {
  return function validateNotNull(value: mixed) {
    if (value === null) {
      return errorString;
    } else {
      return true;
    }
  };
}

export function notUndefined(
  { errorString = "not filled" }: { errorString?: string } = {},
) {
  return function validateNotUndefined(value: mixed) {
    if (value === undefined) {
      return errorString;
    } else {
      return true;
    }
  };
}

export function notEmpty(
  { errorString = "not filled" }: { errorString?: string } = {},
) {
  return function validateNotEmpty(value: mixed) {
    if (value === "") {
      return errorString;
    } else {
      return true;
    }
  };
}

export function composeValidation(...validates) {
  return function validateValue(value: mixed, props: Object) {
    return validates.reduce((red, validate) => {
      if (red !== true) return red;
      const validated = validate(value, props);
      if (validated === true) return true;
      else return validated;
    }, true);
  };
}

export function required(
  { errorString = "not filled" }: { errorString?: string } = {},
) {
  return composeValidation(
    notUndefined({ errorString }),
    notNull({ errorString }),
    notEmpty({ errorString }),
  );
}

export function isTrue(
  { errorString = "not true" }: { errorString?: string } = {},
) {
  return function validateTrue(value: mixed) {
    if (value !== true) {
      return errorString;
    } else {
      return true;
    }
  };
}

export function maxLength(
  max: number,
  { errorString = "too long" }: { errorString?: string } = {},
) {
  return function validateMaxLength(value: mixed) {
    if (typeof value !== "string" || value.length > max) {
      return errorString;
    } else {
      return true;
    }
  };
}

export function lessThan(
  accessor1: (value: mixed, props: Object) => mixed | string,
  accessor2: (value: mixed, props: Object) => mixed | string,
  { errorString = "Invalid" }: { errorString?: string } = {},
) {
  return function validateLessThan(value: mixed, props: Object) {
    if (
      (typeof accessor1 !== "function" && typeof accessor1 !== "string") ||
      (typeof accessor2 !== "function" && typeof accessor2 !== "string")
    )
      throw new Error(
        "lessThan validator takes either a function or a stringPath as accessor type",
      );
    const value1 = typeof accessor1 === "function"
      ? accessor1(value, props)
      : fromJS(value).getIn(accessor1.split("."));
    const value2 = typeof accessor2 === "function"
      ? accessor2(value, props)
      : fromJS(value).getIn(accessor2.split("."));
    // $FlowFixMe
    if (value1 > value2) {
      return errorString;
    } else {
      return true;
    }
  };
}
