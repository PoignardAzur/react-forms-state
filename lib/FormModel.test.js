import { validateModel, convertIn, convertOut } from "./FormModel";
import { composeValidation } from "./ValidationComposition";

describe("FormModel", () => {
  describe("convertIn", () => {
    const inValue = {
      user: {
        firstname: "alan",
        lastname: "ah ok",
        account: {
          email: "my email",
          password: "test",
        },
      },
      trip: {
        destination: {
          country: "Australia",
        },
      },
    };

    it("should handle well empty in", () => {
      const jobs = [
        {
          in: "",
          out: "test",
        },
        {
          in: "user.firstname",
          out: "firstname",
        },
      ];
      const subject = convertIn(inValue, jobs);

      expect(subject).toMatchSnapshot();
    });

    it("should handle well empty out", () => {
      const jobs = [
        {
          in: "user.firstname",
          out: "",
        },
        {
          in: "user.lastname",
          out: "lastname",
        },
      ];
      const subject = convertIn(inValue, jobs);

      expect(subject).toMatchSnapshot();
    });

    it("should handle well empty in and out", () => {});
  });

  describe("convertOut", () => {
    const stateValue = {
      user: {
        firstname: "alan",
        lastname: "ah ok",
        account: {
          email: "my email",
          password: "test",
        },
      },
      trip: {
        destination: {
          country: "Australia",
        },
      },
    };

    it("should handle well empty in", () => {});

    it("should handle well empty out", () => {});

    it("should handle well empty in and out", () => {});
  });

  describe("validateModel", () => {
    const stateValue = {
      user: {
        firstname: "alan",
        lastname: "ah ok",
        account: {
          email: "my email",
          password: "test",
        },
      },
      trip: {
        destination: {
          country: "Australia",
        },
      },
    };

    it("passes the props and value to different validators", () => {
      const spy = jest.fn(() => true);
      const jobs = [
        {
          in: "user.firstname",
          out: "user.firstname",
          validate: spy,
        },
      ];
      const props = {};
      const subject = validateModel(jobs)(stateValue, props);

      expect(spy).toHaveBeenCalledWith("alan", props);
    });

    it("returns true when every validator returns true", () => {
      const jobs = [
        {
          in: "user.firstname",
          out: "user.firstname",
          validate: () => true,
        },
        {
          in: "user.account.email",
          out: "user.account.email",
          validate: () => true,
        },
        {
          in: "user.trip.destination.address.zipcode",
          out: "user.trip.destination.address.zipcode",
          validate: () => true,
        },
      ];
      const props = {};
      const subject = validateModel(jobs)(stateValue, props);

      expect(subject).toBe(true);
    });

    it("returns true if one validator returns null", () => {
      const jobs = [
        {
          in: "user.firstname",
          out: "user.firstname",
          validate: () => true,
        },
        {
          in: "user.account.email",
          out: "user.account.email",
          validate: () => null,
        },
        {
          in: "user.lastname",
          out: "user.lastname",
          validate: () => true,
        },
      ];
      const props = {};
      const subject = validateModel(jobs)(stateValue, props);

      expect(subject).toBe(true);
    });

    it("returns true if one validator returns undefined", () => {
      const jobs = [
        {
          in: "user.firstname",
          out: "user.firstname",
          validate: () => true,
        },
        {
          in: "user.lastname",
          out: "user.lastname",
          validate: () => undefined,
        },
        {
          in: "user.account.email",
          out: "user.account.email",
          validate: () => true,
        },
      ];
      const props = {};
      const subject = validateModel(jobs)(stateValue, props);

      expect(subject).toBe(true);
    });

    it("returns an object if one validator returns an error", () => {
      const jobs = [
        {
          in: "user.firstname",
          out: "user.firstname",
          validate: () => true,
        },
        {
          in: "user.lastname",
          out: "user.lastname",
          validate: () => "not a real one",
        },
        {
          in: "user.account.email",
          out: "user.account.email",
          validate: () => true,
        },
      ];
      const props = {};
      const subject = validateModel(jobs)(stateValue, props);

      expect(subject).toMatchSnapshot();
    });

    it("merges objects if several error are returns by validators", () => {
      const jobs = [
        {
          in: "user.firstname",
          out: "user.firstname",
          validate: () => true,
        },
        {
          in: "user.lastname",
          out: "user.lastname",
          validate: () => "not a real one",
        },
        {
          in: "user.account.email",
          out: "user.account.email",
          validate: () => "bad email",
        },
      ];
      const props = {};
      const subject = validateModel(jobs)(stateValue, props);

      expect(subject).toMatchSnapshot();
    });

    it("handles nested validation", () => {
      const jobs = [
        {
          in: "user.firstname",
          out: "user.firstname",
          validate: () => true,
        },
        {
          in: "user.lastname",
          out: "user.lastname",
          validate: () => "not a real one",
        },
        {
          in: "user.account",
          out: "user.account",
          validate: () => "Weird account",
        },
        {
          in: "user.account.email",
          out: "user.account.email",
          validate: () => "bad email",
        },
      ];
      const props = {};
      const subject = validateModel(jobs)(stateValue, props);

      expect(subject).toMatchSnapshot();
    });

    it("handles nested validation with out structure different than in structure", () => {
      const jobs = [
        {
          in: "firstname",
          out: "user.firstname",
          validate: () => true,
        },
        {
          in: "lastname",
          out: "user.lastname",
          validate: () => "not a real one",
        },
        {
          in: "user.account",
          out: "user.account",
          validate: () => "Weird account",
        },
        {
          in: "user.account.email",
          out: "user.account.email",
          validate: () => "bad email",
        },
      ];
      const props = {};
      const subject = validateModel(jobs)(stateValue, props);

      expect(subject).toMatchSnapshot();
    });

    it("does not crash if validate is not defined", () => {
      const jobs = [
        {
          in: "user.firstname",
          out: "user.firstname",
        },
      ];
      const props = {};
      expect(() => validateModel(jobs)(stateValue, props)).not.toThrow();
    });

    it("should handle well empty in", () => {});

    it("should handle well empty out", () => {});

    it("should handle well empty in and out", () => {});

    it("is usable with composeValidation", () => {
      const spy = jest.fn(() => true);
      const jobs = [
        {
          in: "user.firstname",
          out: "user.firstname",
          validate: () => true,
        },
        {
          in: "user.account.email",
          out: "user.account.email",
          validate: spy,
        },
        {
          in: "user.trip.destination.address.zipcode",
          out: "user.trip.destination.address.zipcode",
          validate: () => true,
        },
      ];
      const props = {};
      const subject = composeValidation(validateModel(jobs))(stateValue, props);

      expect(subject).toBe(true);
      expect(spy).toHaveBeenCalledWith("my email", props);
    });
  });
});
