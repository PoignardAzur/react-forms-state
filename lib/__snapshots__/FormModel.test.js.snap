// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormModel convertIn should handle well empty in 1`] = `
Object {
  "firstname": "alan",
  "test": Object {
    "trip": Object {
      "destination": Object {
        "country": "Australia",
      },
    },
    "user": Object {
      "account": Object {
        "email": "my email",
        "password": "test",
      },
      "firstname": "alan",
      "lastname": "ah ok",
    },
  },
}
`;

exports[`FormModel convertIn should handle well empty in and out 1`] = `
Object {
  "lastname": "ah ok",
}
`;

exports[`FormModel convertIn should handle well empty out 1`] = `
Object {
  "lastname": "ah ok",
}
`;

exports[`FormModel validateModel handles nested validation 1`] = `
Object {
  "user": Object {
    "account": "Weird account",
    "lastname": "not a real one",
  },
}
`;

exports[`FormModel validateModel handles nested validation with out structure different than in structure 1`] = `
Object {
  "user": Object {
    "account": "Weird account",
    "lastname": "not a real one",
  },
}
`;

exports[`FormModel validateModel merges objects if several error are returns by validators 1`] = `
Object {
  "user": Object {
    "account": Object {
      "email": "bad email",
    },
    "lastname": "not a real one",
  },
}
`;

exports[`FormModel validateModel returns an object if one validator returns an error 1`] = `
Object {
  "user": Object {
    "lastname": "not a real one",
  },
}
`;
